File: src/server/sync-metrics.ts
import { AsyncLocalStorage } from "node:async_hooks";

export type SyncMetricContext = {
  sessionId?: string;
  actionId?: string;
  action?: string;
  origin?: "WATCH" | "PHONE" | "UNKNOWN";
  transport?: "HTTPS_WATCH_DIRECT" | "PHONE_RELAY" | "HTTPS_PHONE" | "POLLING" | "UNKNOWN";
};

type SyncMetricFields = SyncMetricContext & {
  event: string;
  status?: number;
  retry?: number;
  bootstrap?: boolean;
  replay?: boolean;
  durationMs?: number;
};

const metricContext = new AsyncLocalStorage<SyncMetricContext>();

/**
 * Logging only: intentionally never writes telemetry to Neon and never throws.
 * Fields are limited to correlation and timing metadata; workout payloads/tokens
 * must not be added here.
 */
export function logSyncMetric(fields: SyncMetricFields) {
  try {
    console.info("TRAKNIO_SYNC_METRIC", JSON.stringify({
      ts: new Date().toISOString(),
      ...metricContext.getStore(),
      ...fields,
    }));
  } catch {
    // Observability is strictly best-effort.
  }
}

export async function withSyncMetricContext<T>(context: SyncMetricContext, execute: () => Promise<T>): Promise<T> {
  return metricContext.run(context, execute);
}

export function currentSyncMetricContext() {
  return metricContext.getStore();
}

File: android-private/traknio-android/app/src/main/java/com/traknio/app/SyncMetrics.kt
package com.traknio.app

import android.util.Log
import org.json.JSONObject

/** Logging-only sync telemetry. It must never influence workout control flow. */
object SyncMetrics {
    fun log(event: String, sessionId: String? = null, actionId: String? = null, action: String? = null, transport: String? = null, status: Int? = null) {
        runCatching {
            val json = JSONObject().put("ts", System.currentTimeMillis()).put("event", event).put("origin", "PHONE")
            sessionId?.takeIf { it.isNotBlank() }?.let { json.put("sessionId", it) }
            actionId?.takeIf { it.isNotBlank() }?.let { json.put("actionId", it) }
            action?.takeIf { it.isNotBlank() }?.let { json.put("action", it) }
            transport?.takeIf { it.isNotBlank() }?.let { json.put("transport", it) }
            status?.let { json.put("status", it) }
            Log.i("TRAKNIO_SYNC_METRIC", json.toString())
        }
    }
}
